Make changes to the task to allow for central scheduling

henryrizzi opened the pull request:
## Description
I implemented the changes which were requested in the original pull request, along with the changes which should make it easier for central scheduling.

These are the big changes:
#### Code
* output is now required of k8s job tasks, as this allows checking for job completion
* __init__ has been changed to _init_k8s, as the central scheduler was having trouble invoking the task.
* signal_complete is now required (mainly to give a configurable output location rather than just touching a local file).
#### Configuration Information
* Add information about the new configuration settings for k8s to `configuration.rst` in the docs
#### Tests/Examples
* Update tests and examples to implement new requirements and check behavior. 



## Motivation and Context
Without having an output from the kubernetes job, there's no way to check job completion. This means that the job wouldn't work in a pipeline. 

## Have you tested this? If so, how?
Tests have been updated to check writing of a file on successful job completion

mcapuccini commented:
@henryrizzi it looks very good. But I don't get what we need signal_complete for, the user's K8S job is expected to touch the output file, so there shouldn't be any need for it. Am I missing something?

mcapuccini commented:
@henryrizzi I think we should assume a shared PersistentVolume between the Job's container and Luigi. So if the Job produces a file, then Luigi can check the output.

henryrizzi commented:
@mcapuccini I changed it to signal complete rather than touching an output file since I thought there should be the option for having other kinds of output methods. 
For example, if someone wanted to instead touch a record in a database on job completion, they could implement a custom signal_complete for their different kind of output.

Would it be okay to implement signal complete as a function that writes to an output, but still leave it as a separate function for maximum versatility?

mcapuccini commented:
@henryrizzi I see your point. I didn't think about jobs that don't do writes to the fs, but it is for sure something that people want to do at a certain point. I think it would be cleaner to redefine the `complete` function, rather than touching a sys file (https://github.com/spotify/luigi/blob/master/luigi/task.py#L475). I will do this fix, and I'll come back to you.

mcapuccini commented on luigi/contrib/k8s_job.py:
```diff
@@ -126,46 +138,63 @@ def spec_schema(self):
         else:
             raise RuntimeError("Kubernetes job " + self.uu_name + " failed")
 
+    def signal_complete(self):
```
could you make this function optional? So if I have a distributed FS and a Job the writes the output file, I can avoid to implement this. Maybe just replace the exception with `pass`?

henryrizzi replied:
This has been implemented as pass now.

mcapuccini commented on luigi/contrib/k8s_job.py:
```diff
@@ -186,8 +216,14 @@ def run(self):
         self.__logger.info("Start tracking Kubernetes Job: " + self.uu_name)
         self.__track_job()
+
+    def output(self):
```
I'd like this to be optional as well. Please remove this the `output` method override. Let's say I write into a DB, and I redefine the `complete` method to check that there is a result record in the DB (instead of letting Luigi to check an output file). Then, I should be allowed to not define the output.

henryrizzi replied:
I'll just replace this with a pass as well and leave the pseudo code for what I would implement it as in combination with the potential `signal_complete`.

henryrizzi replied:
And this has been implemented as pass as well. :)

mcapuccini commented on test/contrib/k8s_job_test.py:
```diff
@@ -51,6 +56,15 @@ class SuccessJob(KubernetesJobTask):
         }]
     }
 
+    def signal_complete(self):
```
Please remove it form this test, and test it in a separate test case (SignalCompleteJob).

mcapuccini commented on test/contrib/k8s_job_test.py:
```diff
@@ -51,6 +56,15 @@ class SuccessJob(KubernetesJobTask):
+        with self.output().open('w') as output:
+            output.write('')
+
+    def output(self):
```
Please remove it form this test, and test it in a separate test case (SignalCompleteJob).

mcapuccini commented on test/contrib/k8s_job_test.py:
```diff
@@ -62,22 +76,38 @@ class FailJob(KubernetesJobTask):
         }]
     }
 
+    def signal_complete(self):
```
Please remove it form this test, and test it in a separate test case (SignalCompleteJob).

mcapuccini commented on test/contrib/k8s_job_test.py:
```diff
@@ -62,22 +76,38 @@ class FailJob(KubernetesJobTask):
+        with self.output().open('w') as output:
+            output.write('')
+
+    def output(self):
```
Please remove it form this test, and test it in a separate test case (SignalCompleteJob).

mcapuccini commented on test/contrib/k8s_job_test.py:
```diff
@@ -62,22 +76,38 @@ class FailJob(KubernetesJobTask):
         success = luigi.run(["SuccessJob", "--local-scheduler"])
         self.assertTrue(success)
 
+        # a file should have been touched on job completion
```
Please remove it form this test, and test it in a separate test case (SignalCompleteJob).

henryrizzi commented:
@mcapuccini I'll add tests for the `SignalCompleteJob` and take the parts of the test which check that out. Should have that ready by tomorrow or the day after at the latest. 👍

henryrizzi commented:
I took a slightly different approach for testing now, and I also changed the example to have the `signal_complete` and `output` method commented out as they're now optional via pass.

The new test which I added uses mock to make sure that `signal_complete` is called on successful job completion, but without necessarily trying to run a job. This is helpful because trying to run two jobs in a local scheduler from the same script will spawn at the same pid, causing a luigi interface error.

If you have any more comments or fixes, please let me know!

mcapuccini commented:
@henryrizzi it looks awesome! Thanks!
